File: ios/calorietracker/Models/HeartRateEntry.swift
import Foundation

enum HeartRateSource: String, Codable, CaseIterable, Sendable {
    case camera
    case manual

    var displayName: String {
        switch self {
        case .camera:
            String(localized: "Camera")
        case .manual:
            String(localized: "Manual")
        }
    }
}

/// A single wellness-oriented spot heart-rate reading. Camera frames and the
/// raw pulse waveform are deliberately never part of this persisted model.
struct HeartRateEntry: Identifiable, Codable, Equatable, Sendable {
    let id: UUID
    let date: Date
    let bpm: Int
    let source: HeartRateSource
    /// Optional normalized confidence from 0...1 for camera measurements.
    /// Manual entries intentionally leave this nil.
    let quality: Double?

    init(
        id: UUID = UUID(),
        date: Date = .now,
        bpm: Int,
        source: HeartRateSource,
        quality: Double? = nil
    ) {
        self.id = id
        self.date = date
        self.bpm = bpm
        self.source = source
        self.quality = quality.map { min(max($0, 0), 1) }
    }
}

File: android/app/src/main/java/com/apoorvdarshan/calorietracker/models/HeartRateEntry.kt
package com.apoorvdarshan.calorietracker.models

import kotlinx.serialization.Serializable
import java.time.Instant
import java.util.UUID

@Serializable
enum class HeartRateSource {
    CAMERA,
    MANUAL
}

/** A single local wellness heart-rate spot check. Camera frames and waveforms are never stored. */
@Serializable
data class HeartRateEntry(
    @Serializable(with = UuidSerializer::class)
    val id: UUID = UUID.randomUUID(),
    @Serializable(with = InstantSerializer::class)
    val date: Instant = Instant.now(),
    val bpm: Int,
    val source: HeartRateSource,
    /** Normalized 0–1 camera signal confidence. Manual entries intentionally leave this null. */
    val quality: Double? = null
) {
    val isValid: Boolean
        get() = bpm in MIN_PLAUSIBLE_BPM..MAX_PLAUSIBLE_BPM &&
            (quality == null || (quality.isFinite() && quality in 0.0..1.0))

    companion object {
        const val MIN_PLAUSIBLE_BPM = 30
        const val MAX_PLAUSIBLE_BPM = 250
    }
}
